### PR TITLE
use tf.shape

### PR DESCRIPTION
**Pull request by emailweixu**
use tf.shape(tensor) instead of tensor.shape so that the operators involving shape can handle dynamic  batch size.
tensor.shape is some python data structure which does not have defined value (i.e. it may contain None) if the shape is dynamically set. We should always use tf.shape() to access the shape of a tensor.

### CODE REVIEW

File: alf/drivers/off_policy_driver.py [hunk 1]
```diff
@@ -292,13 +281,15 @@ def _update(self, experience, weight):
                 lambda state: state[0, ...], experience.state)
         else:
             first_train_state = initial_train_state
-        num_steps = experience.step_type.shape[0]
+        num_steps = tf.shape(experience.step_type)[0]
 
         def create_ta(s):
+            # TensorArray cannot use Tensor (batch_size) as element_shape
+            ta_batch_size = experience.step_type.shape[1]
```
Comment (hnyu) [inline alf/drivers/off_policy_driver.py]: Can we directly do int(batch_size) here?

File: alf/utils/common.py [hunk 2]
```diff
@@ -214,23 +214,39 @@ def _tensor_parameters_only(params):
...
     Args:
         x (Tensor): source tensor
         y (Tensor): target tensor. Only its shape will be used.
-    Returns
+    Returns:
         x with extra singular dimensions.
     """
     assert len(x.shape) <= len(y.shape)
-    assert x.shape == y.shape[:len(x.shape)]
+    tf.assert_equal(tf.shape(x), tf.shape(y)[:len(x.shape)])
```
Comment (hnyu) [inline alf/utils/common.py]: x.shape

Comment (emailweixu) [reply alf/utils/common.py]: What’s the problem?

Comment (emailweixu) [reply alf/utils/common.py]: What’s the problem?

File: alf/drivers/off_policy_driver.py (hunk 1 above)
Comment (emailweixu) [reply alf/drivers/off_policy_driver.py]: I didn’t try. But I guess not.

Comment (witwolf) [reply alf/drivers/off_policy_driver.py]: > Can we directly do int(batch_size) here?

can not convert a tensor to an integer in graph mode

File: alf/drivers/off_policy_driver.py [hunk 3]
```diff
@@ -253,37 +249,30 @@ def _make_time_major(nest):
         for u in tf.range(num_updates):
             if mini_batch_size < batch_size:
```
Comment (witwolf) [inline alf/drivers/off_policy_driver.py]: conditional expression `mini_batch_size < batch_size`  is a tensor  as `batch_size` is a tensor
Peharps we should avoid using a tensor as a bool value

Comment (emailweixu) [reply alf/drivers/off_policy_driver.py]: What is the problem of that?

Comment (witwolf) [reply alf/drivers/off_policy_driver.py]: Somehow  that using a tensor as bool doesn't work in graph mode sometimes

Comment (emailweixu) [reply alf/drivers/off_policy_driver.py]: But this one is not a problem.

**Pull request merged**